proposals/5/README.md

```diff
@@ -0,0 +1,103 @@
+****---
+title: Decomposition of Jenkins X
+linktitle: Decomposition of Jenkins X
+description: Decomposition of Jenkins X
+type: docs
+weight: 50
+---
+
+## 1. Overview
+
+This document outlines the work for decomposing Jenkins X into a series of reusable libraries.
+
+## 1.1 Motivation
+
+At the moment, the source code that comprises the main JX binary is tightly coupled and difficult to work on.  Changes have
+unintended consequences and this makes the codebase fragile.
+
+One of the design goals in Accelerate is to have a **Loosely Coupled Architecture** - in particular: 
+
+_The goal is for your architecture to support the ability of teams to get their work done—from design through to deployment—without requiring high-bandwidth communication between teams._
+
+## 1.2 Background
+
+The main [JX repository](https://github.com/jenkins-x/jx) contains over 250k lines of code, creating Apps / Extensions / Binary Plugins is difficult without adding a dependency back onto the JX repo.
+
+It should be possible to separate the repository out into a series of reusable libraries so only what is needed can be imported.
+
+## 2. Design
+
+There are currently a number of proposed repositories/modules, some of these are currently in flight.
+
+### Existing Modules
+
+* [jx-api](https://github.com/jenkins-x/jx-api): JX CRDs / Installation Requirements
+* [jx-kube-client](https://github.com/jenkins-x/jx-kube-client): A helper module to create a Kubernetes rest config
+* [jx-vault-client](https://github.com/jenkins-x/jx-vault-client): longer term, it may make sense to deprecate this.
+* [go-scm](https://github.com/jenkins-x/go-scm): All 
+
+### Proposed Modules
+
+* jx-kube: It may make sense to refactor some of the higher level utils/kube functions within JX into its own package to make them reusable.
+* Version Streams: Would these make sense to be moved out of JX
+* Gitter?
+* Updatebot (jx create pr)? This would have value outside of the core JX codebase 
+* Storage APIs?
```

Comment:
I would be very careful here. Similar to the _Gitter_ interface, I don't think that extracting these APIs are they are is really a good idea. The APIs they offer are at least partly questionable. Extracting them and moving forward with them will just spread a bad API. 

In general, I think we should be conservative in the number of libraries we extract. Initially, we should aim for the minimal number of libraries to get us of `jx`. Each additional library comes with overhead and I am afraid we go from one extreme (everything in one code base) to the next (huge set of micro libraries). 
We already have such a tremendous amount of repositories (libraries, charts, app) which makes it very hard to know what goes where and how. Execution aside, the idea of `jx` to have many things in one repository has advantages. It is easier to develop and easier to refactor and make changes across the whole codebase. 
There would also be the option to have multiple modules from within the same repository.

Given that we were not able to refactor `jx` in a way that we can use it in a more sensible way, I do agree that extracting libraries make sense to cut the compile-time dependency to the `jx` codebase. Extracting the API (CRDs, PipelineConfig and eventually the InstallRequirements) makes sense. This makes sense also for users trying to extend Jenkins X via Apps. 
Beyond that, I feel we should be careful. Several "API"s are imo not worth extracting. The APIs are flawed and one would just make it even harder to refactor them and get off them. In these cases, I would rather see the start of a new library which a better-designed API and an attempt to move the code over to the new API.